master/internal/db/cache.go

```diff
@@ -0,0 +1,187 @@
+package db
+
+import (
+	"context"
+	"io/fs"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/determined-ai/determined/master/pkg/archive"
+	"github.com/determined-ai/determined/proto/pkg/experimentv1"
+)
+
+type modelDefFolder struct {
+	fileTree   []*experimentv1.FileNode
+	cachedTime time.Time
+	lock       sync.RWMutex
+	path       string
+}
+
+// FileCache is metadata for files cached at file system.
+type FileCache struct {
+	rootDir string
+	maxAge  time.Duration
+	caches  map[int]*modelDefFolder
+	lock    sync.Mutex
+}
+
+// NewFileCache initialize FileCache obj.
+func NewFileCache(rootDir string, maxAge time.Duration) *FileCache {
+	return &FileCache{
+		rootDir: rootDir,
+		maxAge:  maxAge,
+		caches:  make(map[int]*modelDefFolder),
+	}
+}
+
+func (f *FileCache) getOrCreateFolder(expID int) (*modelDefFolder, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	value, ok := f.caches[expID]
+	if !ok {
+		exp := struct {
+			ModelDefinition []byte
+		}{}
+		err := Bun().NewSelect().TableExpr(
+			"experiments").Column("model_definition").Where("id = ?", expID).Scan(context.TODO(), &exp)
+		if err != nil {
+			return nil, err
+		}
+		var fileTree []*experimentv1.FileNode
+		arc, err := archive.FromTarGz(exp.ModelDefinition)
+		if err != nil {
+			return nil, err
+		}
+		err = os.MkdirAll(f.genPath(expID, ""), fs.ModePerm)
+		if err != nil {
+			return nil, err
+		}
+		for _, ar := range arc {
+			if ar.IsDir() {
+				err = os.MkdirAll(f.genPath(expID, ar.Path), fs.ModePerm)
+			} else {
+				err = os.WriteFile(f.genPath(expID, ar.Path), ar.Content, fs.ModePerm)
+			}
+			if err != nil {
+				return nil, err
+			}
+			fileTree = append(fileTree, &experimentv1.FileNode{
+				Path:          ar.Path,
+				ModifiedTime:  timestamppb.New(ar.ModifiedTime.Time),
+				ContentLength: int32(len(ar.Content)),
+				IsDir:         ar.IsDir(),
+				ContentType:   http.DetectContentType(ar.Content),
+				Name:          filepath.Base(ar.Path),
+			})
+		}
+		value = &modelDefFolder{
+			path:       f.genPath(expID, ""),
+			fileTree:   fileTree,
+			cachedTime: time.Now(),
+		}
+		f.caches[expID] = value
+		f.prune()
+	}
+	return value, nil
+}
+
+func (f *FileCache) prune() {
```

Comment:
this prune code does not seem to take into account any files which were present on disk before the master started up.  Effectively, when the master restarts, any files on disk have been "leaked" and will never be reused or cleaned up.

Reply:
When we call `GetModelDefCache` for the first time, I did `os.RemoveAll(rootDir)` so that file system and memory can sync up, but it's true that after initialization, if something tempers with the folder, `prune` does not take into account that.

Reply:
Yes absolutely, I should have done that